apps/prairielearn/src/pages/instructorFileEditor/instructorFileEditor.ts

```diff
@@ -277,120 +243,117 @@ function getHash(contents: string) {
+    sql.soft_delete_file_edit,
+    { user_id, course_id, dir_name, file_name },
+    IdSchema,
+  );
+  debug(`Deleted ${deletedFileEdits.length} previously saved drafts`);
+  for (const file_id of deletedFileEdits) {
+    if (fileEdit?.file_id != null && idsEqual(file_id, fileEdit.file_id)) {
+      debug(`Defer removal of file_id=${file_id} from file store until after reading contents`);
     } else {
-      debug(`Remove file_id=${row.file_id} from file store`);
-      await deleteFile(row.file_id, fileEdit.userID);
+      debug(`Remove file_id=${file_id} from file store`);
+      await deleteFile(file_id, authn_user_id);
     }
   }
+  if (fileEdit == null) return null;
 
-  if (fileEdit.editID && fileEdit.fileID) {
+  let contents: string | undefined;
+  let hash: string | undefined;
+  if (fileEdit.file_id != null) {
     debug('Read contents of file edit');
-    const result = await getFile(fileEdit.fileID);
-    const contents = b64EncodeUnicode(result.contents.toString('utf8'));
-    fileEdit.editContents = contents;
-    fileEdit.editHash = getHash(fileEdit.editContents);
+    const result = await getFile(fileEdit.file_id);
+    contents = b64EncodeUnicode(result.contents.toString('utf8'));
+    hash = getHash(contents);
 
-    debug(`Remove file_id=${fileEdit.fileID} from file store`);
-    await deleteFile(fileEdit.fileID, fileEdit.userID);
+    debug(`Remove file_id=${fileEdit.file_id} from file store`);
+    await deleteFile(fileEdit.file_id, authn_user_id);
   }
+
+  return { fileEdit, contents, hash };
 }
 
 async function updateJobSequenceId(edit_id: string, job_sequence_id: string) {
-  await sqldb.queryAsync(sql.update_job_sequence_id, {
-    id: edit_id,
-    job_sequence_id,
-  });
+  await sqldb.queryAsync(sql.update_job_sequence_id, { id: edit_id, job_sequence_id });
   debug(`Update file edit id=${edit_id}: job_sequence_id=${job_sequence_id}`);
 }
 
-async function writeDraftEdit(fileEdit: {
-  userID: string;
-  authnUserID: string;
-  courseID: string;
-  dirName: string;
-  fileName: string;
-  origHash: string;
-  coursePath: string;
-  uid: string;
-  user_name: string;
+async function writeDraftEdit({
+  user_id,
+  authn_user_id,
+  course_id,
+  dir_name,
+  file_name,
+  orig_hash,
+  editContents,
+}: {
+  user_id: string;
+  authn_user_id: string;
+  course_id: string;
+  dir_name: string;
+  file_name: string;
+  orig_hash: string;
   editContents: string;
 }) {
-  const deletedFileEdits = await sqldb.queryAsync(sql.soft_delete_file_edit, {
-    user_id: fileEdit.userID,
-    course_id: fileEdit.courseID,
-    dir_name: fileEdit.dirName,
-    file_name: fileEdit.fileName,
-  });
-  debug(`Deleted ${deletedFileEdits.rowCount} previously saved drafts`);
-  for (const row of deletedFileEdits.rows) {
-    debug(`Remove file_id=${row.file_id} from file store`);
-    await deleteFile(row.file_id, fileEdit.userID);
+  const deletedFileEdits = await sqldb.queryRows(
+    sql.soft_delete_file_edit,
```

Comment:
Almost certainly not a question for this PR, but: AFAICT there's no reason this should be a soft-delete table. We read from this table in exactly one place and we always use `WHERE deleted_at IS NULL`.

Reply:
I was about to say that this could be a problem since the table includes a `file_id`, which also has soft-deletes, but this column is not an official foreign key, so there would be no cascade delete, so that is not actually a problem. In any case, as you mentioned, probably an issue for a separate PR.